atd-toolbox/s3_restore_cr3_pdfs.py

```diff
@@ -0,0 +1,270 @@
+
+HEADERS = {
+          "X-Hasura-Admin-Secret": HASURA_ADMIN_KEY,
+          "Content-Type": "application/json",
+          }
+
+bucket = 'atd-vision-zero-editor'
+
+
+# https://github.com/cityofaustin/atd-airflow/blob/master/dags/python_scripts/atd_vzd_cr3_scan_pdf_records.py#L24
+def is_valid_metadata(metadata: dict) -> bool:
+    if metadata.get("last_update", None) is not None \
+       and metadata.get("file_size", None) is not None \
+       and metadata.get("mime_type", None) is not None \
+       and metadata.get("encoding", None) is not None:
+        return True
+    return False
+
+
+# setup and parse arguments
+try:
+    argparse = argparse.ArgumentParser(description = 'Utility to restore last valid PDF in S3 for ATD VZ')
+
+    argparse.add_argument("-p", "--production",
+            help = 'Specify the use of production environment',
+            action = 'store_true')
+
+    argparse.add_argument("-c", "--crashes",
+            help = 'Specify JSON file containing crashes to operate on. Format: { "crashes": [ crash_id_0, crash_id_1, .. ] }',
+            required=True,
+            metavar = 'crashes.json')
+
+    argparse.add_argument("--i-understand",
+            help = 'Do not ask the user to acknoledge that this program changes the state of S3 objects and the database.',
+            action = 'store_true')
+
+    args = argparse.parse_args()
+except:
+    # a stderr log is not needed, argparse croaks verbosly
+    sys.exit(1)
+
+
+# This program will change the state of S3 objects.  Make sure the user is OK with what is about to happen.
+try:
+    if not args.i_understand:
+        print('')
+        print("Warning: This program changes S3 Objects.")
+        print('')
+        print("This program will restore previous file versions which are larger than 10K for crashes specified in the JSON object you provide.")
+        print("If there is a 'application/pdf' stored as a previous version of a specified crash's CR3, this program will restore that file to the current version.")
+        print("This program will also update the databases cr3_file_metadata field for crash records based on the S3 file restored.")
+        print("Please type 'I understand' to continue.")
+        print('')
+        ack = input()
+        assert(re.match("^i understand$", ack, re.I))
+except:
+    eprint("User acknoledgement failed.")
+    sys.exit(1)
+
+
+# verify that environment AWS variables were available and have populated values to be used to auth to AWS
+try:
+    assert(ACCESS_KEY is not None and SECRET_KEY is not None)
+except:
+    eprint("Please set environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY")
+    sys.exit(1)
+
+
+# verify that environment Hasura variables were available and have populated values
+try:
+    assert(HASURA_ADMIN_KEY is not None and HASURA_ENDPOINT is not None)
+except:
+    eprint("Please set environment variables HASURA_ADMIN_KEY and HASURA_ENDPOINT")
+    sys.exit(1)
```

Comment:
It never occurred to me we can use the assert function like this (outside of a test) 🥇